api-mock/fetcher.ts

```diff
@@ -0,0 +1,248 @@
+                headers,
+                body: input ? JSON.stringify(input) : undefined,
+            });
+            if (expectedStatus !== null && response.status !== expectedStatus) {
+                throw new Error(`Expected status ${expectedStatus} but received ${response.status}`);
+            }
+
+            if (response.ok === false) {
+                throw {
+                    status: response.status,
+                    body: await response.json(),
+                };
+            }
+            const body = await response.json().catch(() => undefined);
+            const text = await response.text().catch(() => JSON.stringify(body));
+
+            const responseHeaders: Record<string, string | undefined> = {};
+
+            response.headers.forEach((_, key) => {
+                responseHeaders[key] = response.headers.get?.(key) || undefined;
+            });
+
+            callback?.(null, {
+                body,
+                status: response.status,
+                headers: responseHeaders,
+                text,
+            });
+        } catch (error) {
+            callback?.(error, null);
+        }
+    }
+
+    function post(pathValue: string): ChainedRequest {
+        path = pathValue;
+        method = "post";
+        return this;
+    }
+
+    function get(pathValue: string): ChainedRequest {
+        path = pathValue;
+        method = "get";
+        return this;
+    }
+
+    function set(header: string, value: string): ChainedRequest {
+        headers[header] = value;
+        return this;
+    }
+
+    function send(data: any): ChainedRequest {
+        input = data;
+        return this;
+    }
+
+    function expect(status: number, cb?: Callback): ChainedRequest {
+        expectedStatus = status;
+        if (cb !== undefined) {
+            this.end(cb);
+        }
+        return this;
+    }
+
+    function end(cb: Callback) {
+        callback = cb;
+        executeRequest();
+    }
+
+    return {
+        post,
+        get,
+        set,
+        send,
+        expect,
+        end,
+    };
+}
+
+export async function queryAPI({
+    method,
+    path,
+    input,
+    headers,
+    returnResponse,
+    skipInit,
+}: {
+    method: "post" | "get" | "delete" | "put";
+    path: string;
+    input?: object;
+    headers?: Record<string, string>;
+    returnResponse?: boolean;
+    skipInit?: boolean;
+}) {
+    if (!skipInit && status === "NOT_READY") {
+        await initApp();
+    }
+    try {
+        let response = await fetch(`http://localhost:${apiMockPort}${path}`, {
+            method,
+            headers: {
+                "Content-Type": "application/json",
+                ...headers,
+            },
+            body: JSON.stringify(input),
+        });
+
+        if (returnResponse) {
+            return response;
+        }
+
+        if (!response.ok) {
+            throw response;
+        }
+
+        return await response.json().catch(() => undefined);
```

Comment:
Are there cases where we expect non-json responses? If not, we should just let this throw.

Reply:
@sandrodesouza I think this is the only remaining comment on this.

Reply:
yes, there is this case here EmailVerificationRecipe.getInstance().updateSessionIfRequiredPostEmailVerification

Reply:
that also seems to return through `res.json`

Reply:
yes but could be undefined

Reply:
I see, I thought you wrapped those in objects, but I guess this works as well.